R/tags.R

```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
 #'   use a relative path (the base path being the Shiny application directory),
 #'   not an absolute path.
 #'
+#' @note `includeHTML()` directly inlines the contents of the file, which is only
+#'   suitable for fragments of HTML. To include a full HTML document, such as
+#'   the generated output of an HTML R Markdown or Quarto document, use
+#'   `includeHTMLDocument()`.
+#'
 #' @rdname include
 #' @name include
 #' @aliases includeHTML
 #' @export
 includeHTML <- function(path) {
   lines <- readLines(path, warn=FALSE, encoding='UTF-8')
+
+  if (detect_html_document(lines)) {
+    rlang::warn(c(
+      "`includeHTML()` was provided a `path` that appears to be a complete HTML document.",
+      "x" = paste("Path:", path),
+      "i" = "Use `includeHTMLDocument()` to include an HTML document."
+    ))
+  }
+
   return(HTML(paste8(lines, collapse='\n')))
 }
 
+detect_html_document <- function(lines) {
+  if (length(lines) > 1) {
+    lines <- paste8(lines, collapse = "\n")
+  }
+  lines <- trimws(lines)
+
+  # A complete html document starts with doctype declaration or opening <html>
+  if (!grepl("^(<!DOCTYPE html>|<html)", lines, ignore.case = TRUE)) {
+    return(FALSE)
+  }
+  # and ends by closing the `</html>` tag
+  if (!grepl("</html>$", lines, ignore.case = TRUE)) {
+    return(FALSE)
+  }
+
+  # There are more requirements for the HTML document to be technically complete
+  # and valid, but the above conditions are sufficient for us to warn that the
+  # document should not be treated as an HTML fragment.
+  TRUE
+}
+
+#' @note `includeHTMLDocument()` embeds the contents of the file in an
```

Comment:
What is the difference between `@note` and `@describeIn`? Could we use `@describeIn include DESC` here?

Reply:
I guess this would have to be applied to the whole file.

No need to change it, still curious about the difference though.